internal/tar_interpreter.go

```diff
@@ -45,42 +45,56 @@ func (tarInterpreter *FileTarInterpreter) unwrapRegularFile(fileReader io.Reader
 		}
 	}
 	fileDescription, haveFileDescription := tarInterpreter.Sentinel.Files[fileInfo.Name]
+	isIncremented := haveFileDescription && fileDescription.IsIncremented
+	// todo: clearer catchup backup detection logic
+	isCatchup := tarInterpreter.createNewIncrementalFiles
 
-	// If this file is incremental we use it's base version from incremental path
-	if haveFileDescription && tarInterpreter.Sentinel.IsIncremental() && fileDescription.IsIncremented {
-		err := ApplyFileIncrement(targetPath, fileReader, tarInterpreter.createNewIncrementalFiles)
-		return errors.Wrapf(err, "Interpret: failed to apply increment for '%s'", targetPath)
+	if localFileInfo, _ := getLocalFileInfo(targetPath); localFileInfo != nil {
+		isPageFile := isPagedFile(localFileInfo, targetPath)
+		return unwrapToExistFile(fileReader, fileInfo, targetPath, isPageFile, isIncremented, isCatchup)
 	}
-	err := PrepareDirs(fileInfo.Name, targetPath)
+	return unwrapToNewFile(fileReader, fileInfo, targetPath, isIncremented)
+}
+
+// unwrap the file from tar to existing local file
+func unwrapToExistFile(fileReader io.Reader, fileInfo *tar.Header, targetPath string,
+	isPageFile, isIncremented, isCatchup bool) error {
```

Comment:
You pass 3 bool arguments to a function. It seems like "if-based" logic :) Sometime it means that it is possible to create a base class and create several embedding for using instead of check bool flags.